### PR TITLE
Implements dim, dot, matmul, equal

### LINKED ISSUE #37: Implement Default dim Functionality for Base Tensor Type
**User Story A:** As a Data Scientist using Syft's Base Tensor type, we want to implement a default method for computing operations on a Tensor of arbitrary type. For this ticket to be complete, diag() should return a list of the dimensions of the tensor. For a reference on the operation this performs check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation.

**Acceptance Criteria:**
- If the Base Tensor type's attribute "encrypted" is set to True, it should return a NotImplemented error.
- a unit test demonstrating the correct operation on the Base Tensor type implemented over int and float Tensors.
- inline documentation in the python code. For inspiration on inline documentation, please check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation for this operator.

### LINKED ISSUE #38: Implement Default dot Functionality for Base Tensor Type
**User Story A:** As a Data Scientist using Syft's Base Tensor type, we want to implement a default method for computing operations on a Tensor of arbitrary type. For this ticket to be complete, dot() should return a new tensor. For a reference on the operation this performs check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation.

**Acceptance Criteria:**
- If the Base Tensor type's attribute "encrypted" is set to True, it should return a NotImplemented error.
- a unit test demonstrating the correct operation on the Base Tensor type implemented over int and float Tensors.
- inline documentation in the python code. For inspiration on inline documentation, please check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation for this operator.

### LINKED ISSUE #40: Implement Default equal Functionality for Base Tensor Type
**User Story A:** As a Data Scientist using Syft's Base Tensor type, we want to implement a default method for computing operations on a Tensor of arbitrary type. For this ticket to be complete, equal() should return a new tensor. For a reference on the operation this performs check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation.

**Acceptance Criteria:**
- If the Base Tensor type's attribute "encrypted" is set to True, it should return a NotImplemented error.
- a unit test demonstrating the correct operation on the Base Tensor type implemented over int and float Tensors.
- inline documentation in the python code. For inspiration on inline documentation, please check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation for this operator.

### LINKED ISSUE #61: Implement Default matmul Functionality for Base Tensor Type
**User Story A:** As a Data Scientist using Syft's Base Tensor type, we want to implement a default method for computing operations on a Tensor of arbitrary type. For this ticket to be complete, matmul() should return a new tensor. For a reference on the operation this performs check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation.

**Acceptance Criteria:**
- If the Base Tensor type's attribute "encrypted" is set to True, it should return a NotImplemented error.
- a unit test demonstrating the correct operation on the Base Tensor type implemented over int and float Tensors.
- inline documentation in the python code. For inspiration on inline documentation, please check out [PyTorch](http://pytorch.org/docs/master/tensors.html)'s documentation for this operator.

### PR DESCRIPTION
**Pull request by samsontmr**
Fixes #37 
Fixes #38 
Fixes #61 
Fixes #40

### CODE REVIEW

File: syft/tensor.py [hunk 1]
```diff
@@ -50,6 +87,17 @@ def __isub__(self, arr_like):
...
 
+
+    def __dot__(self, arr_like):
+        """Returns inner product of two tensors."""
+        return syft.dot(self, arr_like)
+
+
+    def __matmul__(self, arr_like):
+        """Performs matrix multiplication between two tensors"""
+        return syft.matmul(self, arr_like)
```
Comment (TheOriginalAlex) [inline syft/tensor.py]: For `__dot__` and `__matmul__`, I'd run `_ensure_ndarray` on the `arr_like` to make sure the new array being passed is a numpy ndarray.

Comment (samsontmr) [inline syft/tensor.py]: Hm so does `_ensure_ndarray` converts a tensorbase object into a numpy array?

Comment (TheOriginalAlex) [inline syft/tensor.py]: Yea.  I suppose it should be changed to _ensure_tensorbase.

File: syft/__init__.py [hunk 2]
```diff
@@ -1 +1,2 @@
 from .tensor import TensorBase
+from .tensor import *
```
Comment (erfannoury) [inline syft/__init__.py]: I think it's not a good practice to have this kind of imports in Python. It is always better to be explicit about what you are importing.

Comment (samsontmr) [inline syft/__init__.py]: Yes, was hoping to get feedback on how this is normally implemented. Do you normally add every method to init?

Comment (erfannoury) [inline syft/__init__.py]: You can either do `from .tensor import x, y, z`, or if the number of methods is large, you can just do `import .tensor` and then do use `tensor.x` in code.

File: syft/tensor.py [hunk 3]
```diff
@@ -1,88 +1,182 @@
...
+    addition, subtraction, multiplication and division, and also dot and matrix products.
     """
 
     def __init__(self, arr_like, encrypted=False):
         self.data = _ensure_ndarray(arr_like)
         self.encrypted = encrypted
 
-    def __add__(self, arr_like):
-        """Performs element-wise addition between two array like objects"""
+    def __add__(self, tensor):
```
Comment (erfannoury) [inline syft/tensor.py]: I think it would be better if you also add `__radd__` and `__rsub__`.

File: syft/__init__.py (hunk 2 above)
Comment (TheOriginalAlex) [inline syft/__init__.py]: Also, I think the standalone functions like `syft.matmul`, etc. should go in a separate file.  Like this PyTorch example: https://github.com/pytorch/pytorch/blob/master/torch/functional.py

File: tests/test_tensor.py [hunk 4]
```diff
@@ -1,80 +1,165 @@
...
+        t1 = TensorBase(np.array([1, 2, 3]))
+        t2 = TensorBase(np.array([1, 4, 5]))
+        self.assertTrue(t1 != t2)
+
+
+class DotTests(unittest.TestCase):
+    def testDotInt(self):
+        t1 = TensorBase(np.array([1, 2, 3]))
+        t2 = TensorBase(np.array([4, 5, 6]))
+        self.assertTrue(syft.dot(t1, t2), 32)
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 5]
```diff
@@ -1,80 +1,165 @@
...
+
+    def testDotFloat(self):
+        t1 = TensorBase(np.array([1.3, 2.5, 3.7]))
+        t2 = TensorBase(np.array([4.9, 5.8, 6.5]))
+        self.assertTrue(syft.dot(t1, t2), 43.07)
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 6]
```diff
@@ -1,80 +1,165 @@
...
+
+
+class MatmulTests(unittest.TestCase):
+    def testMatmul1DInt(self):
+        t1 = TensorBase(np.array([1, 2, 3]))
+        t2 = TensorBase(np.array([4, 5, 6]))
+        self.assertTrue(syft.matmul(t1, t2), syft.dot(t1, t2))
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 7]
```diff
@@ -1,80 +1,165 @@
...
+
+    def testMatmul1DFloat(self):
+        t1 = TensorBase(np.array([1.3, 2.5, 3.7]))
+        t2 = TensorBase(np.array([4.9, 5.8, 6.5]))
+        self.assertTrue(syft.matmul(t1, t2), syft.dot(t1, t2))
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 8]
```diff
@@ -1,80 +1,165 @@
...
+
+    def testMatmul2DIdentity(self):
+        t1 = TensorBase(np.array([[1, 0],
+                                  [0, 1]]))
+        t2 = TensorBase(np.array([[5.8, 6.5],
+                                  [7.8, 8.9]]))
+        self.assertTrue(syft.matmul(t1, t2), [[5.8, 6.5],
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 9]
```diff
@@ -1,80 +1,165 @@
...
+                                              [7.8, 8.9]])
+
+    def testMatmul2DInt(self):
+        t1 = TensorBase(np.array([[1.3, 2.5],
+                                  [3.4, 4.5]]))
+        t2 = TensorBase(np.array([[5.8, 6.5],
+                                  [7.8, 8.9]]))
+        self.assertTrue(syft.matmul(t1, t2), [[19, 22],
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: tests/test_tensor.py [hunk 10]
```diff
@@ -1,80 +1,165 @@
...
+                                              [43, 50]])
+
+    def testMatmul2DFloat(self):
+        t1 = TensorBase(np.array([[1.3, 2.5],
+                                  [3.4, 4.5]]))
+        t2 = TensorBase(np.array([[5.8, 6.5],
+                                  [7.8, 8.9]]))
+        self.assertTrue(syft.matmul(t1, t2), [[27.04, 8.45],
```
Comment (erfannoury) [inline tests/test_tensor.py]: Shouldn't this be `self.assertEqual(...)`?

File: syft/__init__.py (hunk 2 above)
Comment (samsontmr) [inline syft/__init__.py]: good idea. Should we call it math.py?

Comment (samsontmr) [inline syft/__init__.py]: @TheOriginalAlex @erfannoury what do you think of using __all__ in combination with `from .tensor import *`? I see many libraries doing this.

Comment (TheOriginalAlex) [inline syft/__init__.py]: Yea.  `math.py` sounds good.  :)

File: syft/tensor.py (hunk 3 above)
Comment (samsontmr) [inline syft/tensor.py]: This should probably be left to a different issue.

File: tests/test_tensor.py (hunk 4 above)
Comment (samsontmr) [inline tests/test_tensor.py]: oh yes, good catch!

File: syft/__init__.py (hunk 2 above)
Comment (TheOriginalAlex) [inline syft/__init__.py]: Yea.  I've seen it done multiple places in PyTorch and other libraries.  I think that's a good way of doing it when we have a bunch of functions and we wanna export the contents of a module without exporting the module itself.

Comment (samsontmr) [inline syft/__init__.py]: Should we keep equal in tensor.py?

Comment (TheOriginalAlex) [inline syft/__init__.py]: Sure.  Along with the _ensure* functions.

**Comment by iamtrask** [comment]
Forgive me if it's written elsewhere, but is there a general philosophy between which operations should be defined in BaseTensor and which should be defined within math.py?

**Comment by samsontmr** [comment]
We decided to keep the math operations in math and the others (like equal()) in tensor.py. `__operations__` have to be in tensor.py since they are defined under the TensorBase class.

**Comment by iamtrask** [comment]
Works for me! Great thought. :)

**Pull request merged**